backend/src/main/java/fk/prof/backend/aggregator/CpuSamplingAggregationBucket.java

```diff
@@ -0,0 +1,83 @@
+package fk.prof.backend.aggregator;
+
+import fk.prof.backend.exception.AggregationFailure;
+import fk.prof.backend.model.request.RecordedProfileIndexes;
+import fk.prof.aggregation.MethodIdLookup;
+import fk.prof.aggregation.cpusampling.CpuSamplingFrameNode;
+import fk.prof.aggregation.cpusampling.CpuSamplingTraceDetail;
+import recording.Recorder;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class CpuSamplingAggregationBucket {
+  private final MethodIdLookup methodIdLookup = new MethodIdLookup();
+  private final ConcurrentHashMap<String, CpuSamplingTraceDetail> traceDetailLookup = new ConcurrentHashMap<>();
+
+  public Set<String> getAvailableTraces() {
+    return traceDetailLookup.keySet();
+  }
+
+  public CpuSamplingTraceDetail getTraceDetail(String traceName) {
+    return traceDetailLookup.get(traceName);
+  }
+
+  /**
+   * Generates a unmodifiable view of method id -> method name lookup
+   * This method should be called once the aggregation window, this map is member of, has expired.
+   * Calling this method while window is in progress and receiving profiles for aggregation will result in incomplete method name lookup being generated
+   *
+   * @return returns methodId -> method name lookup map
+   */
+  public Map<Long, String> getMethodNameLookup() {
+    return this.methodIdLookup.generateReverseLookup();
+  }
+
+  /**
+   * Aggregates stack samples in the bucket. Throws {@link AggregationFailure} if aggregation fails
+   *
+   * @param stackSampleWse
+   */
+  public void aggregate(Recorder.StackSampleWse stackSampleWse, RecordedProfileIndexes indexes)
+      throws AggregationFailure {
+
+    for (Recorder.StackSample stackSample : stackSampleWse.getStackSampleList()) {
+      //NOTE: As of now, only those samples are aggregated which are associated with a trace id
+      if (stackSample.hasTraceId()) {
+        String trace = indexes.getTrace(stackSample.getTraceId());
+        if (trace == null) {
+          throw new AggregationFailure("Unknown trace id encountered in stack sample, aborting aggregation of this profile");
+        }
+        CpuSamplingTraceDetail traceDetail = traceDetailLookup.computeIfAbsent(trace,
+            key -> new CpuSamplingTraceDetail()
+        );
+
+        List<Recorder.Frame> frames = stackSample.getFrameList();
+        if (frames != null && frames.size() > 0) {
```

Comment:
Is null check relevant? This is a repeated field, protobuf doesn't allow it to be null afaik.

Let us not add unnecessary defense, because it makes code hard to read and debug (as it silently zips by failures in unexpected ways).

We want things to fail with exceptions so we notice them when unexpected things happen.